sources/corepackUtils.ts

```diff
@@ -132,6 +132,75 @@ function isValidBinSpec(x: unknown): x is BinSpec {
   return typeof x === `object` && x !== null && !Array.isArray(x) && Object.keys(x).length > 0;
 }
 
+async function download(installTarget: string, url: string, algo: string, binPath: string | null = null): Promise<DownloadSpec> {
+  // Creating a temporary folder inside the install folder means that we
+  // are sure it'll be in the same drive as the destination, so we can
+  // just move it there atomically once we are done
+
+  const tmpFolder = folderUtils.getTemporaryFolder(installTarget);
+  debugUtils.log(`Downloading to ${tmpFolder}`);
+
+  const stream = await httpUtils.fetchUrlStream(url);
+
+  const parsedUrl = new URL(url);
+  const ext = path.posix.extname(parsedUrl.pathname);
+
+  let outputFile: string | null = null;
+  let sendTo: any;
+
+  if (ext === `.tgz`) {
+    const {default: tar} = await import(`tar`);
+    sendTo = tar.x({
+      strip: 1,
+      cwd: tmpFolder,
+      filter: binPath ? path => {
+        const pos = path.indexOf(`/`);
+        if (pos === -1 || path.slice(pos + 1) !== binPath)
+          return false;
+
+        return true;
+      } : undefined,
+    });
+  } else if (ext === `.js`) {
+    outputFile = path.join(tmpFolder, path.posix.basename(parsedUrl.pathname));
+    sendTo = fs.createWriteStream(outputFile);
+  }
+  stream.pipe(sendTo);
+
+  const streamHash = stream.pipe(createHash(algo));
+  await once(sendTo, `finish`);
+
+  if (binPath) {
+    const downloadedBin = path.join(tmpFolder, binPath);
+    outputFile = path.join(tmpFolder, path.basename(downloadedBin));
+    try {
+      await renameSafe(downloadedBin, outputFile);
+    } catch (err) {
+      if ((err as nodeUtils.NodeError)?.code === `ENOENT`)
+        throw new Error(`Cannot locate '${binPath}' in downloaded tarball`, {cause: err});
+
+      throw err;
+    }
+
+    // Calculate the hash of the bin file
+    const fileStream = fs.createReadStream(outputFile);
+    const fileHash = fileStream.pipe(createHash(algo));
+    await once(fileStream, `close`);
+
+    return {
+      tmpFolder,
+      outputFile,
+      hash: fileHash.digest(`hex`),
+    };
+  }
+
+  return {
+    tmpFolder,
+    outputFile,
+    hash: streamHash.digest(`hex`),
+  };
```

Comment:
Given those return are almost exactly similar, it might make more sense to use only one, wdyt?

Reply:
okay